content/_code-samples/key-derivation/key-derivation.py

```diff
@@ -0,0 +1,327 @@
+        """
+        return base58.b58encode_check(XRPL_SEED_PREFIX + self.bytes).decode()
+
+    def encode_hex(self):
+        """
+        Returns a string representation of this seed as hexadecimal.
+        """
+        return self.bytes.hex().upper()
+
+    def encode_rfc1751(self, correct_rfc1751=None):
+        """
+        Returns a string representation of this seed as an RFC-1751 encoded
+        passphrase.
+        """
+        # Use the default byte order swap this Seed was generated with
+        # unless the method call overrides it.
+        if correct_rfc1751 is None:
+            correct_rfc1751=self.correct_rfc1751
+
+        if correct_rfc1751:
+            buf = self.bytes
+        else:
+            buf = swap_byte_order(self.bytes)
+        return RFC1751.key_to_english(buf)
+
+    @property
+    def ed25519_secret_key(self):
+        """
+        Returns a 32-byte Ed25519 secret key (bytes).
+        Saves the calculation for later calls.
+        """
+        if self._ed25519_sec is None:
+            self._ed25519_sec = sha512half(self.bytes)
+        return self._ed25519_sec
+
+    @property
+    def ed25519_public_key(self):
+        """
+        33-byte Ed25519 public key (bytes)—really a 32-byte key
+        prefixed with the byte 0xED to indicate that it's an Ed25519 key.
+        """
+        if self._ed25519_pub is None:
+            self._ed25519_pub = (ED_PREFIX +
+                                 ed25519.publickey(self.ed25519_secret_key))
+        return self._ed25519_pub
+
+    @property
+    def secp256k1_secret_key(self):
+        """
+        32-byte secp256k1 secret key (bytes)
+        """
+        if self._secp256k1_sec is None:
+            self.derive_secp256k1_master_keys()
+        return self._secp256k1_sec
+
+    @property
+    def secp256k1_public_key(self):
+        """
+        33-byte secp256k1 account public key (bytes)
+        """
+        if self._secp256k1_pub is None:
+            self.derive_secp256k1_master_keys()
+        return self._secp256k1_pub
+
+    @property
+    def secp256k1_root_public_key(self):
+        """
+        33-byte secp256k1 root public key (bytes)
+        This is the public key used for validators.
+        """
+        if self._secp256k1_root_pub is None:
+            self.derive_secp256k1_master_keys()
+        return self._secp256k1_root_pub
+
+    def derive_secp256k1_master_keys(self):
+        """
+        Uses the XRPL's convoluted key derivation process to get the
+        secp256k1 master keypair for this seed value.
+        Saves the values to the object for later reference.
+        """
+
+        root_sec_i = secp256k1_secret_key_from(self.bytes)
+        root_pub_point = keys.get_public_key(root_sec_i, curve.secp256k1)
+        root_pub_b = compress_secp256k1_public(root_pub_point)
+        fam_b = bytes(4) # Account families are unused; just 4 bytes of zeroes
+        inter_pk_i = secp256k1_secret_key_from(root_pub_b+fam_b)
```

Comment:
I don't love that we're using `+` for string concatenation in a math-heavy function like this. I'd either:
1) Add a comment that we are concatenating bytes here, not adding points or
2) Use something like. `b''.join([root_pub_b, fam_b])` instead of `root_pub_b + fam_b`